[GraphQL] Changes to support inline upload File

leobispo opened the pull request:
Checking if it is multipart document extracting the data from multipart form field, parsing it and assigning the multipart files to the correct graphql variables.

leobispo commented:
Hi all,

The code is completed, but I still don't have any idea how to proceed with tests. I couldn't understand how to use the ApolloTestsServer to test my code.

Looking forward for some comments

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -264,6 +270,67 @@ private void handlePostQuery(RoutingContext rc, GraphQLQuery query, String opera
+          idx = Integer.parseInt(path[pathSize - 1]);
+          --pathSize;
+        } catch (final Exception e) {
+          idx = -1;
```
Don't use exceptions to find if something isn't a number, it's very expensive.

leobispo replied:
Done

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -193,6 +196,9 @@ private void handlePost(RoutingContext rc, Buffer body) {
         handlePostJson(rc, body, rc.queryParams().get("operationName"), variables);
         break;
+      case "multipart/form-data":
+        handlePostMultipart(rc, rc.queryParams().get("operationName"), variables);
```
This feature should be disabled by default. Can you a new entry to the handler options and check its value here?

leobispo replied:
Done

leobispo commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -304,8 +371,13 @@ private void executeOne(RoutingContext rc, GraphQLQuery query) {
+    }
+
+    contentType = contentType.toLowerCase();
+    return contentType.contains("multipart/form-data") ? "multipart/form-data" : contentType;
```
No. the header will return the boundary as well. I could substring it... but I don't know
if it will make sense.

tsegismont replied:
See my comment above. Master has an implementation of `getContentType` that works with multipart requests too.

tsegismont replied:
>Unless vertx is doing something under the hood (and I doubt it), we must remove the boundary from check

I actually does. See the latest implementation in master branch that uses `parsedHeaders`

tsegismont replied:
Please do not change the `getContentType` implementation that is currently in `master`. It works fine because it uses the `parsedHeaders` values form `vertx-web`.

leobispo commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -304,8 +371,13 @@ private void executeOne(RoutingContext rc, GraphQLQuery query) {
   private String getContentType(RoutingContext rc) {
-    String contentType = rc.parsedHeaders().contentType().value();
-    return contentType.isEmpty() ? "application/json" : contentType.toLowerCase();
+    String contentType = rc.request().headers().get(HttpHeaders.CONTENT_TYPE);
```
Done...

tsegismont replied:
The PR has been rebased on `master`, but I was suggesting to stick to the current `getContentType` implementation. Just tried it and `MultipartRequestTest` works fine.

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -198,6 +207,67 @@ private void handlePostQuery(RoutingContext rc, GraphQLQuery query, String opera
+    handlePostQuery(rc, (GraphQLQuery) graphQLInput, operationName, variables);
+  }
+
+  private GraphQLInput parseApolloAttributes(RoutingContext rc) {
```
Can you copy the relevant part of the spec in the method documentation?

https://github.com/jaydenseric/graphql-multipart-request-spec

tsegismont replied:
Can you also change the name as this is not Apollo stuff but rather a GraphQL community effort to standardize uploads?

tsegismont commented on vertx-web-graphql/src/test/java/io/vertx/ext/web/handler/graphql/MultipartRequestTest.java:
```diff
@@ -0,0 +1,111 @@
+  }
+}
+
+public class MultipartRequestTest extends WebTestBase {
```
Can you add a test for:

- same file used in different variables
- multiple files
- multiple operations (batch)

tsegismont commented on vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/GraphQLHandlerImpl.java:
```diff
@@ -198,6 +207,67 @@ private void handlePostQuery(RoutingContext rc, GraphQLQuery query, String opera
+    final MultiMap attrs = rc.request().formAttributes();
+    final Map<String, List<String>> map = Json.decodeValue(attrs.get("map"),
+        new TypeReference<HashMap<String, List<String>>>() {});
+    final GraphQLQuery input = Json.decodeValue(attrs.get("operations"), GraphQLQuery.class);
```
According to the spec, the `operations` attribute can be a single request (object) or batch (array)

leobispo commented:
@tsegismont Can you check it One more time.

Now we have batch calls, few more tests and I updated the method documentation.

Please let me know it you need any other change.

tsegismont commented:
@leobispo can you please update the documentation to describe the feature? It's not necessary to include code on how to setup the Apollo client code for file upload. Pointing to external documentation is fine.

Also, could you add a JS integration test? Thank you

leobispo commented:
@tsegismont Can you check it One more time.

- I fixed some incompatibilities with 4.0.0
- Removed the contentType changes
- Wrote the integration test
- Wrote a short documentation on how to configure the feature.

Please let me know it you need any other change.

tsegismont commented:
@leolux I will have a look asap. Thank you

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,54 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
 
+=== GraphQL multipart request
+
+To use https://github.com/jaydenseric/graphql-multipart-request-spec which will allow you to upload one or more files, using a single mutation call.
```
Can you please reword this introduction sentence? Not sure what you mean.

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,54 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+
+[source,$lang]
+----
+GraphQLHandlerOptions graphQLHandlerOptions = new GraphQLHandlerOptions()
```
You should put the code samples in the `Examples` class. When the Vert.x stack is built, they will be translated to the other supported languages.

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,54 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+  .setRequestMultipartEnabled(true);
+----
+
+Next, configure the router with the `BodyHandler`. This handler will parse the multipart HTTP body.
```
When using Vert.x class names in the doc, put them in a Javadoc link reference. Then the doc will have a link to the Javadoc:

```suggestion
Next, configure the router with the {@link io.vertx.ext.web.handler.BodyHandler}. This handler will parse the multipart HTTP body.
```

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,54 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+
+[source,$lang]
+----
+GraphQLScalarType uploadScalar = new GraphQLScalarType.Builder()
```
Reading this, it seems to me we should have this scalar type in the code and only document how to declare it.

tsegismont commented on vertx-web-graphql/tests/apollo/package.json:
```diff
@@ -12,6 +12,9 @@
     "unfetch": "4.1.0",
     "apollo-link-ws": "1.0.18",
-    "subscriptions-transport-ws": "0.9.16"
+    "subscriptions-transport-ws": "0.9.16",
```
Can you change to fixed versions please?

leobispo commented on vertx-web-graphql/tests/apollo/apollo.test.js:
```diff
@@ -4,11 +4,20 @@ import {HttpLink} from 'apollo-link-http';
 const wsUri = 'ws://localhost:8080/graphql';
 
+const client = new ApolloClient({
+  cache: new InMemoryCache(),
```
This is to use the ApolloClient itself. Without the memory cache the ApolloClient will throw an error (it is mandatory to start it).

If you want, I can try to use only the method that is implementing the multipart call

leobispo commented:
Hi @tsegismont 

Tnx for the comments! (Sorry for the delay)

I fixed all the comments :).

Please let me know if you have any more comments.

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,42 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+
+https://github.com/jaydenseric/graphql-multipart-request-spec[GraphQL multipart request] is an interoperable multipart form field structure
+for `GraphQL` requests. By enabling this functionality, `GraphQL clients` will be able to upload files using a single mutation call. All the
+server side file handling will be abstracted by the {@link io.vertx.ext.web.handler.graphql.GraphQLHandler}.
```
```suggestion
All the server-side file handling will be abstracted by the {@link io.vertx.ext.web.handler.graphql.GraphQLHandler}.
```

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,42 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+{@link examples.GraphQLExamples#setupGraphQLHandlerMultipart}
+----
+
+IMPORTANT: If the router does not have a BodyHandler, the multipart request parser will not be able to handle the `GraphQL` mutation call.
```
```suggestion
IMPORTANT: If the router does not have a {@link io.vertx.ext.web.handler.BodyHandler}, the multipart request parser will not be able to handle the GraphQL mutation call.
```

tsegismont commented on vertx-web-graphql/src/main/asciidoc/index.adoc:
```diff
@@ -199,6 +199,42 @@ Here you can find how to configure the Apollo SubscriptionClient: https://github
+
+IMPORTANT: If the router does not have a BodyHandler, the multipart request parser will not be able to handle the `GraphQL` mutation call.
+
+Finally, create the Upload scalar and set it to the `RuntimeWiring`:
```
```suggestion
Finally, create the `Upload` scalar and set it to the `RuntimeWiring`:
```

tsegismont commented on vertx-web-graphql/src/main/java/examples/GraphQLExamples.java:
```diff
@@ -80,11 +82,37 @@ public void handlerSetupBatching(GraphQL graphQL) {
+    Router router = Router.router(Vertx.vertx());
+
+    router.route().handler(BodyHandler.create());
+    router.route("/graphql").order(100).handler(graphQLHandler);
```
```suggestion
    router.route("/graphql").handler(graphQLHandler);
```

tsegismont commented on vertx-web-graphql/src/main/java/examples/GraphQLExamples.java:
```diff
@@ -80,11 +82,37 @@ public void handlerSetupBatching(GraphQL graphQL) {
+    router.route("/graphql").order(100).handler(graphQLHandler);
+  }
+
+  public void setGraphqlMultipart() {
```
I believe this example can be removed. It's already shown in `setupGraphQLHandlerMultipart`. So removing this and rewording the doc should be enough.

leobispo commented:
@tsegismont 
All the comments added to the doc.

Let me know it you need any other changes :)

tsegismont commented:
@leobispo great work :clap:  Thank you very much!
